File: cplus/leetcode/202503/algo1027.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 - 合法子序列 DP
class Solution {
public:
    int longestArithSeqLength(vector<int>& nums) {
        // 自行解答
        unordered_map<long, int> dp;
        int ans = 2;
        int mx = ranges::max(nums), mn = ranges::min(nums);

        for (auto& x: nums) {
            for (int d=(mn-mx)/2; d<=(mx-mn)/2; d++) {
                long mask = (long) d << 32 | x;
                long prev = (long) d << 32 | x - d;
                dp[mask] = dp[prev] + 1;
                ans = max(ans, dp[mask]);
            }
        }
        return ans;

    }
};

File: python/algo/202503/algo1027.py
from collections import defaultdict
from typing import List

# 动态规划 - 合法子序列 DP
class Solution:
    def longestArithSeqLength(self, nums: List[int]) -> int:
        # 自行解答， 3688ms 刚刚通过
        dp = defaultdict(int)
        ans = 2 # 至少为 2
        mx, mn = max(nums), min(nums)

        for x in nums:
            for d in range((mn-mx)//2, (mx-mn)//2+1):  # mx - mn 差的一半
                dp[(d, x)] = dp[(d, x-d)] + 1
                if dp[(d,x)] > ans: ans = dp[(d,x)]
        
        return ans